cirq/ops/fsim_gate.py

```diff
@@ -164,3 +165,268 @@ def __repr__(self) -> str:
+        c = cmath.exp(-1j * self.phi)
+        f1 = cmath.exp(1j * self.delta_plus + 1j * self.delta_minus)
+        f2 = cmath.exp(1j * self.delta_plus -
+                       1j * self.delta_minus_off_diagonal)
+        f3 = cmath.exp(1j * self.delta_plus +
+                       1j * self.delta_minus_off_diagonal)
+        f4 = cmath.exp(1j * self.delta_plus - 1j * self.delta_minus)
+        f5 = cmath.exp(2j * self.delta_plus)
+        return np.array([
+            [1, 0, 0, 0],
+            [0, f1 * a, f2 * b, 0],
+            [0, f3 * b, f4 * a, 0],
+            [0, 0, 0, f5 * c],
+        ])
+
+    def _resolve_parameters_(self, param_resolver: 'cirq.ParamResolver'
+                            ) -> 'cirq.PhasedFSimGate':
+        return PhasedFSimGate(
+            protocols.resolve_parameters(self.theta, param_resolver),
+            protocols.resolve_parameters(self.phi, param_resolver),
+            protocols.resolve_parameters(self.delta_plus, param_resolver),
+            protocols.resolve_parameters(self.delta_minus, param_resolver),
+            protocols.resolve_parameters(self.delta_minus_off_diagonal,
+                                         param_resolver))
+
+    def _apply_unitary_(self,
+                        args: 'cirq.ApplyUnitaryArgs') -> Optional[np.ndarray]:
+        if cirq.is_parameterized(self):
+            return None
+        oi = args.subspace_index(0b01)
+        io = args.subspace_index(0b10)
+        ii = args.subspace_index(0b11)
+        if (self.theta != 0 or self.delta_minus != 0 or
+                self.delta_minus_off_diagonal != 0):
+            rx = protocols.unitary(cirq.rx(2 * self.theta))
+            rz1 = protocols.unitary(
+                cirq.rz(self.delta_minus - self.delta_minus_off_diagonal))
+            rz2 = protocols.unitary(
+                cirq.rz(self.delta_minus + self.delta_minus_off_diagonal))
+            inner_matrix = rz1 @ rx @ rz2
+            out = cirq.apply_matrix_to_slices(args.target_tensor,
+                                              inner_matrix,
+                                              slices=[oi, io],
+                                              out=args.available_buffer)
+        else:
+            out = args.target_tensor
+        if self.phi != 0:
+            out[ii] *= cmath.exp(-1j * self.phi)
+        if self.delta_plus != 0:
+            f = cmath.exp(1j * self.delta_plus)
+            out[oi] *= f
+            out[io] *= f
+            out[ii] *= f * f
+        return out
+
+    def _decompose_(self, qubits) -> 'cirq.OP_TREE':
+        q0, q1 = qubits
+        q0_z_exponent_before = self.phase_angles_before[0] / np.pi
+        q1_z_exponent_before = self.phase_angles_before[1] / np.pi
+        q0_z_exponent_after = self.phase_angles_after[0] / np.pi
+        q1_z_exponent_after = self.phase_angles_after[1] / np.pi
+        yield cirq.Z(q0)**q0_z_exponent_before
+        yield cirq.Z(q1)**q1_z_exponent_before
```

Comment:
I'd suggest writing this in terms of rotations, and can also use names that correspond with the docstring:
```python
b0, b1 = self.phase_angles_before
a0, a1 = self.phase_angles_after
yield cirq.rz(q0, b0)
yield cirq.rz(q1, b1)
yield FSimGate(self.theta, self.phi).on(q0, q1)
yield cirq.rz(q0, a0)
yield cirq.rz(a1, a1)
```

Reply:
This is nearly identical to my original implementation. Unfortunately, protocol consistency checks are sensitive to global phase, so we have to give up `__unitary__()[0, 0] == 1` or we have to yield `cirq.Z` instead of `cirq.rz` here. I chose the latter so that `__unitary__` agrees with eq (43) from [QS supplement](https://arxiv.org/abs/1910.11333).

(Per other comments, I'll be updating this PR to agree with eq (18) in [FH paper](https://arxiv.org/abs/2010.07965), but this merely renames existing "delta" parameters, so I think I'll keep the constraint that `__unitary__()[0, 0] == 1`.)

Reply:
(We could also yield a global phase gate, but that seems even uglier.)

Reply:
I see. In that case, perhaps include a comment why this differs from the docstring, which shows `Rz` gates. In addition, would be good to refactor to avoid valuating each angle property twice, and you might also consider following the `rz` [implementation](https://github.com/quantumlib/Cirq/blob/master/cirq/ops/common_gates.py#L1332) which uses `sympy.pi` where appropriate to preserve symbolic values.

Reply:
Good suggestions. All done.